[BUG] Detector deletion fails if underlying monitor deletion fails or is not found

```
[2023-03-17T16:40:05,368][ERROR][o.o.a.u.AlertingException] [612b1994a4037b8727a21580fde13874] Alerting error: OpenSearchStatusException[Monitor with fIxu7YYBiwgZNOohwZfl is not found]
[2023-03-17T16:40:05,370][ERROR][o.o.a.u.AlertingException] [612b1994a4037b8727a21580fde13874] Alerting error: java.lang.NullPointerException
[2023-03-17T16:40:05,370][ERROR][o.o.s.u.SecurityAnalyticsException] [612b1994a4037b8727a21580fde13874] Security Analytics error:
[2023-03-17T16:40:05,370][ERROR][o.o.s.u.SecurityAnalyticsException] [612b1994a4037b8727a21580fde13874] Security Analytics error:
AlertingException[Monitor with fIxu7YYBiwgZNOohwZfl is not found]; nested: Exception[org.opensearch.OpenSearchStatusException: Monitor with fIxu7YYBiwgZNOohwZfl is not found];
        at org.opensearch.alerting.util.AlertingException$Companion.wrap(AlertingException.kt:70)
        at org.opensearch.alerting.transport.TransportDeleteMonitorAction$DeleteMonitorHandler.getMonitor(TransportDeleteMonitorAction.kt:132)
        at org.opensearch.alerting.transport.TransportDeleteMonitorAction$DeleteMonitorHandler.access$getMonitor(TransportDeleteMonitorAction.kt:96)
        at org.opensearch.alerting.transport.TransportDeleteMonitorAction$DeleteMonitorHandler$getMonitor$1.invokeSuspend(TransportDeleteMonitorAction.kt)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:233)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)
[2023-03-17T16:40:05,370][WARN ][r.suppressed             ] [612b1994a4037b8727a21580fde13874] path: /_plugins/_security_analytics/detectors/rIxu7YYBiwgZNOoh0JdX, params: {detector_id=rIxu7YYBiwgZNOoh0JdX}
SecurityAnalyticsException[Monitor with fIxu7YYBiwgZNOohwZfl is not found]; nested: Exception[org.opensearch.alerting.util.AlertingException: Monitor with fIxu7YYBiwgZNOohwZfl is not found];
        at org.opensearch.securityanalytics.util.SecurityAnalyticsException.wrap(SecurityAnalyticsException.java:51)
        at org.opensearch.securityanalytics.transport.TransportDeleteDetectorAction$AsyncDeleteDetectorAction.lambda$finishHim$0(TransportDeleteDetectorAction.java:205)
        at org.opensearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:73)
        at org.opensearch.action.ActionRunnable$2.doRun(ActionRunnable.java:88)
        at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:815)
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.Exception: org.opensearch.alerting.util.AlertingException: Monitor with fIxu7YYBiwgZNOohwZfl is not found
        ... 9 more

--- a/src/main/java/org/opensearch/securityanalytics/transport/TransportDeleteDetectorAction.java
+++ b/src/main/java/org/opensearch/securityanalytics/transport/TransportDeleteDetectorAction.java
@@ -156,7 +156,6 @@ public class TransportDeleteDetectorAction extends HandledTransportAction<Delete
 
         private void onGetResponse(Detector detector) {
             List<String> monitorIds = detector.getMonitorIds();
-            String ruleIndex = detector.getRuleIndex();
             ActionListener<DeleteMonitorResponse> deletesListener = new GroupedActionListener<>(new ActionListener<>() {
                 @Override
                 public void onResponse(Collection<DeleteMonitorResponse> responses) {
@@ -176,8 +175,13 @@ public class TransportDeleteDetectorAction extends HandledTransportAction<Delete
 
                 @Override
                 public void onFailure(Exception e) {
-                    if (counter.compareAndSet(false, true)) {
-                        finishHim(null, e);
+                    if(isOnlyMonitorOrIndexMissingExceptionThrownByGroupedActionListener(e, detector.getId())) {
+                        deleteDetectorFromConfig(detector.getId(), request.getRefreshPolicy());
+                    } else {
+                        log.error(String.format(Locale.ROOT, "Failed to delete detector %s", detector.getId()), e);
+                        if (counter.compareAndSet(false, true)) {
+                            finishHim(null, e);
+                        }
                     }
                 }
             }, monitorIds.size());
@@ -231,6 +235,7 @@ public class TransportDeleteDetectorAction extends HandledTransportAction<Delete
         private void finishHim(String detectorId, Exception t) {
             threadPool.executor(ThreadPool.Names.GENERIC).execute(ActionRunnable.supply(listener, () -> {
                 if (t != null) {
+                    log.error(String.format(Locale.ROOT, "Failed to delete detector %s",detectorId), t);
                     if (t instanceof OpenSearchStatusException) {
                         throw t;
                     }
@@ -239,6 +244,30 @@ public class TransportDeleteDetectorAction extends HandledTransportAction<Delete
                     return new DeleteDetectorResponse(detectorId, NO_VERSION, RestStatus.NO_CONTENT);
                 }
             }));
+        }
+
+        private boolean isOnlyMonitorOrIndexMissingExceptionThrownByGroupedActionListener(
+                Exception ex,
+                String detectorId
+        ) {
+            // grouped action listener listens on mutliple listeners but throws only one exception. If multiple
+            // listeners fail the other exceptions are added as suppressed exceptions to the first failure.
+            int len = ex.getSuppressed().length;
+            for (int i = 0; i <= len; i++) {
+                Throwable e = i == len ? ex : ex.getSuppressed()[i];
+                if (e.getMessage().matches("(.*)Monitor(.*) is not found(.*)")
+                                || e.getMessage().contains(
+                                        "Configured indices are not found: [.opendistro-alerting-config]")
+                ) {
+                    log.error(
+                            String.format(Locale.ROOT, "Monitor or jobs index already deleted." +
+                                    " Proceeding with detector %s deletion", detectorId),
+                            e);
+                } else {
+                    return false;
+                }
+            }
+            return true;
         }
     }
 }

--- a/src/test/java/org/opensearch/securityanalytics/SecurityAnalyticsRestTestCase.java
+++ b/src/test/java/org/opensearch/securityanalytics/SecurityAnalyticsRestTestCase.java
@@ -239,6 +239,18 @@ public class SecurityAnalyticsRestTestCase extends OpenSearchRestTestCase {
         return makeRequest(client, "POST", String.format(Locale.getDefault(), "/_plugins/_alerting/monitors/%s/_execute", monitorId), params, null);
     }
 
+    protected Response deleteAlertingMonitorIndex() throws IOException {
+        return makeRequest(client(), "DELETE", String.format(Locale.getDefault(), "/.opendistro-alerting-config"), new HashMap<>(), null);
+    }
+
+    protected Response deleteAlertingMonitor(String monitorId) throws IOException {
+        return deleteAlertingMonitor(client(), monitorId);
+    }
+
+    protected Response deleteAlertingMonitor(RestClient client, String monitorId) throws IOException {
+        return makeRequest(client, "DELETE", String.format(Locale.getDefault(), "/_plugins/_alerting/monitors/%s", monitorId), new HashMap<>(), null);
+    }
+
     protected List<SearchHit> executeSearch(String index, String request) throws IOException {
         return executeSearch(index, request, true);
     }

--- a/src/test/java/org/opensearch/securityanalytics/resthandler/DetectorRestApiIT.java
+++ b/src/test/java/org/opensearch/securityanalytics/resthandler/DetectorRestApiIT.java
@@ -65,6 +65,48 @@ public class DetectorRestApiIT extends SecurityAnalyticsRestTestCase {
         Assert.assertEquals("Create detector failed", RestStatus.CREATED, restStatus(createResponse));
     }
 
+    @SuppressWarnings("unchecked")
+    public void testDeletingADetector_MonitorNotExists() throws IOException {
+        String index = createTestIndex(randomIndex(), windowsIndexMapping());
+
+        // Execute CreateMappingsAction to add alias mapping for index
+        Request createMappingRequest = new Request("POST", SecurityAnalyticsPlugin.MAPPER_BASE_URI);
+        // both req params and req body are supported
+        createMappingRequest.setJsonEntity(
+                "{ \"index_name\":\"" + index + "\"," +
+                        "  \"rule_topic\":\"" + randomDetectorType() + "\", " +
+                        "  \"partial\":true" +
+                        "}"
+        );
+
+        Response response = client().performRequest(createMappingRequest);
+        assertEquals(HttpStatus.SC_OK, response.getStatusLine().getStatusCode());
+        // Create detector #1 of type test_windows
+        Detector detector1 = randomDetectorWithTriggers(getRandomPrePackagedRules(), List.of(new DetectorTrigger(null, "test-trigger", "1", List.of(randomDetectorType()), List.of(), List.of(), List.of(), List.of())));
+        String detectorId1 = createDetector(detector1);
+
+        String request = "{\n" +
+                "   \"query\" : {\n" +
+                "     \"match\":{\n" +
+                "        \"_id\": \"" + detectorId1 + "\"\n" +
+                "     }\n" +
+                "   }\n" +
+                "}";
+        List<SearchHit> hits = executeSearch(Detector.DETECTORS_INDEX, request);
+        SearchHit hit = hits.get(0);
+
+        String monitorId = ((List<String>) ((Map<String, Object>) hit.getSourceAsMap().get("detector")).get("monitor_id")).get(0);
+
+        Response deleteMonitorResponse = deleteAlertingMonitor(monitorId);
+        assertEquals(200, deleteMonitorResponse.getStatusLine().getStatusCode());
+        entityAsMap(deleteMonitorResponse);
+
+        Response deleteResponse = makeRequest(client(), "DELETE", SecurityAnalyticsPlugin.DETECTOR_BASE_URI + "/" + detectorId1, Collections.emptyMap(), null);
+        Assert.assertEquals("Delete detector failed", RestStatus.OK, restStatus(deleteResponse));
+        hits = executeSearch(Detector.DETECTORS_INDEX, request);
+        Assert.assertEquals(0, hits.size());
+    }
+
         @SuppressWarnings("unchecked")
     public void testCreatingADetector() throws IOException {
         String index = createTestIndex(randomIndex(), windowsIndexMapping());

